ui/raidboss/data/06-ew/raid/p2n.ts

```diff
@@ -0,0 +1,178 @@
+import NetRegexes from '../../../../../resources/netregexes';
+import Outputs from '../../../../../resources/outputs';
+import { callOverlayHandler } from '../../../../../resources/overlay_plugin_api';
+import { Responses } from '../../../../../resources/responses';
+import ZoneId from '../../../../../resources/zone_id';
+import { RaidbossData } from '../../../../../types/data';
+import { PluginCombatantState } from '../../../../../types/event';
+import { TriggerSet } from '../../../../../types/trigger';
+
+export interface Data extends RaidbossData {
+  bodyActor?: PluginCombatantState;
+}
+
+const triggerSet: TriggerSet<Data> = {
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+  zoneId: ZoneId.AsphodelosTheSecondCircle,
+  timelineFile: 'p2n.txt',
+  triggers: [
+    {
+      id: 'P2N Murky Depths',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '680F', source: 'Hippokampos', capture: false }),
+      response: Responses.aoe(),
+    },
+    {
+      id: 'P2N Doubled Impact',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '680E', source: 'Hippokampos' }),
+      response: Responses.sharedTankBuster(),
+    },
+    {
+      id: 'P2N Spoken Cataract',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: ['67F8', '67F7', '67F9'], source: 'Hippokampos' }),
+      delaySeconds: 1,
+      promise: async (data) => {
+        const callData = await callOverlayHandler({
+          call: 'getCombatants',
+        });
+        if (!callData || !callData.combatants || !callData.combatants.length) {
+          console.error('SpokenCataract: failed to get combatants: ${JSON.stringify(callData)}');
+          return;
+        }
+        // This is the real hippo, according to hp.
+        const hippos = callData.combatants.filter((c) => c.BNpcID === 13721);
+        if (hippos.length !== 1) {
+          console.error('SpokenCataract: There is not exactly one Hippo?!?: ${JSON.stringify(hippos)}');
+          data.bodyActor = undefined;
+          return;
+        }
+        data.bodyActor = hippos[0];
+      },
+      alertText: (data, matches, output) => {
+        if (!data.bodyActor) {
+          console.error('SpokenCataract: No boss actor found. Did the promise fail?');
+          return;
+        }
+        const heading = data.bodyActor?.Heading;
+        let bodyHeading = null;
+
+        if (heading < -3.0 || heading > 3.0)
```

Comment:
Rather than an if-else tower, could you do something like https://github.com/quisquous/cactbot/blob/main/ui/raidboss/data/05-shb/eureka/delubrum_reginae_savage.ts#L255?

In other words something more like:
```
// N = 0, E = 1, etc
const heading =  Math.round(2 - 2 * parseFloat(data.bodyActor.Heading) / Math.PI) % 4;
```

The reasoning here is (1) I think this is a lot more condensed, (2) probably this sort of thing should be a utility function because we've used something like it enough and so using the same formula will make it easier to consolidate them later and (3) I think it's better to just round fully to closest direction in case things are slightly off more than the 0.1 radians that you've given a tolerance for.

Reply:
My geometry is (also...) a little weak so I haven't come up with such an elegant formula but if you say it works it is much better indeed.

> I think it's better to just round fully to closest direction in case things are slightly off more than the 0.1 radians that you've given a tolerance for.

Yes and no, had I started with this it would have been an even harder to debug error that boss headings were off 90° sometimes (because closest at time of log line emission wasn't always closest to actual spell direction) while they are fine the rest of the time.

Reply:
> My geometry is (also...) a little weak so I haven't come up with such an elegant formula but if you say it works it is much better indeed.

You don't have to trust me! You can put in some numbers into a javascript console and try out the formula and see if it would give you the same answers you already have.

> closest at time of log line emission wasn't always closest to actual spell direction

Yeah, log lines sometimes have stale positions.  It's very frustrating.

Reply:
Fixed in a8562bbece06d2fd3a33b1cbac9473feb52ee86b